lib/shapely/impl/src/singleton.rs

```diff
@@ -0,0 +1,143 @@
+//! This module defines helpers for defining singletons and associated enum types. A singleton is
+//! a type with one possible value. It is used mainly for a type level programming purposes.
+
+/// Defines singleton types. For the following input:
```

Comment:
Hmm, I know only one definition of singleton, and this is not "type without data"

Reply:
This one? https://cs.brynmawr.edu/~rae/papers/2012/singletons/paper.pdf :D